File: src/control_turtlesim/src/move_straight.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"    // Pub: send move message
#include "turtlesim/Pose.h" // Sub: receive position
#include "iostream"


// Confirm command
// rostopic list
// rostopic info /turtle1/cmd_vel
// rosmsg show geometry_msgs/Twist


void move(ros::Publisher velocity_publisher, double speed, double distance, bool isForward)
{
    geometry_msgs::Twist vel_msg;

    if(isForward) {
        vel_msg.linear.x = abs(speed);
    } else {
        vel_msg.linear.x = -abs(speed);
    }
    
    double t0 = ros::Time::now().toSec();
    double current_distance = 0.0;
    ros::Rate loop_rate(100);
    do {
        velocity_publisher.publish(vel_msg);
        double t1 = ros::Time::now().toSec();
        current_distance = speed * (t1-t0); // s = v.(t1-t0)
        ros::spinOnce();
        loop_rate.sleep();
    } while (current_distance < distance);

    vel_msg.linear.x = 0;
    velocity_publisher.publish(vel_msg);

}

void poseCallback(const turtlesim::Pose::ConstPtr &pose_message){

    float x {pose_message->x};
    float y {pose_message->y};
    float theta {pose_message->theta};

    ROS_INFO("x: %f", x ); // %f :float
    ROS_INFO("y: %f", y );
    ROS_INFO("theta: %f", theta );

}

int main(int argc, char **argv)
{
    // Init new node with name "move_straight_cpp"
    ros::init(argc, argv, "move_straight_cpp");
    ros::NodeHandle n;
    double speed, angular_speed;
    double distance, angle;
    bool isForward, clockwise;

    // Declare Pub v
    ros::Publisher velocity_publisher = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 1000);

    // Declare Sub to show location
    ros::Subscriber pose_subscribe = n.subscribe("/turtle1/pose", 10, poseCallback);

    // Test code
    move(velocity_publisher, 1, 4, false);

    // Wait
    ros::Rate loop_rate(0.5);
    ros::spin();

    return 0;
}

File: src/control_turtlesim/scripts/move_straight.py
#!/usr/bin/env python
from turtle import pos
import rospy
from geometry_msgs.msg import Twist # For PUB
from turtlesim.msg import Pose      # For SUB
import math
import time

# is_forward: True > forward, False > Go back
def move(velocity_publisher, speed, distance, is_forward):

    velocity_message = Twist()

    # Get curent location
    global x, y
    x0 = x # Init location x
    y0 = y # Init location y

    if (is_forward):
        velocity_message.linear.x = abs(speed)
    else:
        velocity_message.linear.x = -abs(speed)

    distance_moved = 0.0
    loop_rate = rospy.Rate(10) # 10Hz Using for sleep

    # Send move message
    while True:
        rospy.loginfo("Turtlesim moves")
        velocity_publisher.publish(velocity_message)
        loop_rate.sleep()
        distance_moved = abs( math.sqrt( (x-x0)**2 + (y-y0)**2 ) )
        print(distance_moved)

        if not (distance_moved < distance):
            rospy.loginfo("Reached")
            break

    # Stop the robot when the distance is moved
    velocity_message.linear.x = 0
    velocity_publisher.publish(velocity_message)

# Sub callback func, get location and set global x, y, yaw
def poseCallback(pose_message):
    global x
    global y, yaw
    x = pose_message.x
    y = pose_message.y
    yaw = pose_message.theta

if __name__ == '__main__':

    try:
        # Init node
        rospy.init_node('turtlesim_motion_pose', anonymous=True)

        # Declare Pub
        velocity_publisher = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)

        # Declare Sub
        pose_subscriber = rospy.Subscriber('/turtle1/pose', Pose, poseCallback)
        time.sleep(2)

        move(velocity_publisher, 1.0, 4.0, True)
        
    except rospy.ROSInterruptException:
        rospy.loginfo("Node terminated")


# Confirm command
# rostopic list
# rostopic info /turtle1/cmd_vel
# rosmsg show geometry_msgs/Twist